File: lambda_calculus/parser.py
import re
from lambda_calculus.syntax import Variable, Abstraction, Application


class ParseError(Exception):
    pass


class Parser:    
    def __init__(self):
        self.lambda_symbol_pattern = r'[λ\\]'
        self.variable_pattern = r'[a-zA-Z][a-zA-Z0-9_]*'
        self.token_pattern = re.compile(
            f'{self.lambda_symbol_pattern}|'
            r'\.|'   
            r'\(|\)|'
            f'{self.variable_pattern}'
        )
    
    def tokenize(self, input_string):
        tokens = self.token_pattern.findall(input_string)
        tokens = [token.strip() for token in tokens if token.strip()]
        return tokens
    
    def parse(self, input_string):
        tokens = self.tokenize(input_string)
        if not tokens:
            raise ParseError("Empty input")
        expr, remaining = self._parse_expr(tokens)
        if remaining:
            raise ParseError(f"Unexpected tokens at end: {' '.join(remaining)}")
        
        return expr
    
    def _parse_expr(self, tokens):
        if not tokens:
            raise ParseError("Unexpected end of input")
        left, tokens = self._parse_atom(tokens)
        while tokens and tokens[0] not in [')', '.']:
            right, tokens = self._parse_atom(tokens)
            left = Application(left, right)
        
        return left, tokens
    
    def _parse_atom(self, tokens):
        if not tokens:
            raise ParseError("Unexpected end of input")
        
        token = tokens[0]
        
        if token == '(':
            _, tokens = tokens[0], tokens[1:]
            expr, tokens = self._parse_expr(tokens)
            if not tokens or tokens[0] != ')':
                raise ParseError("Missing closing parenthesis")
            _, tokens = tokens[0], tokens[1:]
            
            return expr, tokens
        
        elif token == 'λ' or token == '\\':
            _, tokens = tokens[0], tokens[1:]
            if not tokens or not re.match(self.variable_pattern, tokens[0]):
                raise ParseError("Expected variable after lambda")
            
            param, tokens = tokens[0], tokens[1:]
            
            if not tokens or tokens[0] != '.':
                raise ParseError("Expected dot after lambda parameter")
            
            _, tokens = tokens[0], tokens[1:]
            
            body, tokens = self._parse_expr(tokens)
            
            return Abstraction(param, body), tokens
        elif re.match(self.variable_pattern, token):
            return Variable(token), tokens[1:]
        else:
            raise ParseError(f"Unexpected token: {token}")


def parse(input_string):
    parser = Parser()
    return parser.parse(input_string)

File: web/parser.js
class ParseError extends Error {
    constructor(message) {
        super(message);
        this.name = "ParseError";
    }
}

class Parser {
    constructor() {
        this.lambdaSymbolPattern = /[λ\\]/;
        this.variablePattern = /^[a-zA-Z][a-zA-Z0-9_]*$/;
        this.tokenPattern = /[λ\\]|\.|[(]|[)]|[a-zA-Z][a-zA-Z0-9_]*/g;
    }
    
    tokenize(inputString) {
        const tokens = inputString.match(this.tokenPattern) || [];
        return tokens.filter(token => token.trim() !== '');
    }
    
    parse(inputString) {
        const tokens = this.tokenize(inputString);
        if (tokens.length === 0) {
            throw new ParseError("Empty input");
        }
        
        const [expr, remaining] = this._parseExpr(tokens);
        if (remaining.length > 0) {
            throw new ParseError(`Unexpected tokens at end: ${remaining.join(' ')}`);
        }
        
        return expr;
    }
    
    _parseExpr(tokens) {
        if (tokens.length === 0) {
            throw new ParseError("Unexpected end of input");
        }
        
        let [left, remainingTokens] = this._parseAtom(tokens);
        
        while (remainingTokens.length > 0 && ![')', '.'].includes(remainingTokens[0])) {
            const [right, newRemaining] = this._parseAtom(remainingTokens);
            left = new Application(left, right);
            remainingTokens = newRemaining;
        }
        
        return [left, remainingTokens];
    }

    _parseAtom(tokens) {
        if (tokens.length === 0) {
            throw new ParseError("Unexpected end of input");
        }
        
        const token = tokens[0];
        
        if (token === '(') {
            const remainingTokens = tokens.slice(1);
            const [expr, afterExpr] = this._parseExpr(remainingTokens);
            
            if (afterExpr.length === 0 || afterExpr[0] !== ')') {
                throw new ParseError("Missing closing parenthesis");
            }
            return [expr, afterExpr.slice(1)];
        }
        
        else if (token === 'λ' || token === '\\') {
            const afterLambda = tokens.slice(1);
            
            if (afterLambda.length === 0 || !this.variablePattern.test(afterLambda[0])) {
                throw new ParseError("Expected variable after lambda");
            }
            
            const param = afterLambda[0];
            const afterParam = afterLambda.slice(1);
            
            if (afterParam.length === 0 || afterParam[0] !== '.') {
                throw new ParseError("Expected dot after lambda parameter");
            }
            
            const afterDot = afterParam.slice(1);
            const [body, afterBody] = this._parseExpr(afterDot);
            
            return [new Abstraction(param, body), afterBody];
        }
        
        else if (this.variablePattern.test(token)) {
            return [new Variable(token), tokens.slice(1)];
        }
        else {
            throw new ParseError(`Unexpected token: ${token}`);
        }
    }
}

function parse(inputString) {
    const parser = new Parser();
    return parser.parse(inputString);
}
